### PR TITLE
Fix several bugs in fv_regional_bc.F90 relating to uninitialized or incorrectly initialized memory.

### LINKED ISSUE #218: uninitialized or incorrectly initialized memory in fv_regional_bc.F90
**Describe the bug**

The ps_reg variable in fv_regional_bc.F90 is initialized to -9999999 Pa, which propagates throughout the model. All other boundary arrays were initialized to signalling NaN. Not initializing this one to signalling NaN masked many other bugs. 

These are some of the bugs it masked inside the GFDL_atmos_cubed_sphere repository; there are others outside it:

1. Remapping winds does a two point average of the current point and -9999999.
2. Remapping winds uses -9999999 as pressure on velocity boundary points beyond the scalar boundary points.
3. The -9999999 is copied to Atm%ps throughout the domain, beyond the boundary, on all processors that touch the domain boundary.
4. Due to item 3, the RRFS hasn't been able to run in DEBUG=ON mode since this bug was introduced.
5. The pressure data is not read in on velocity points outside the scalar boundary.  (It can't be, since the the boundary condition files lack that data entirely.)
6. The blending code blends data beyond boundary regions.
7. The code accesses i-1 and j-1 indices, but processes the south boundary before the east and west boundary. That means two south boundary processes are accessing uninitialized regions instead of the east and west boundary.

There are other issues, outside the dynamical core, which were masked by this bug. I'll link to those issues once I create or find them.

- https://github.com/ufs-community/ufs-weather-model/issues/1222
- https://github.com/ufs-community/ufs-weather-model/issues/1436
- https://github.com/ufs-community/ccpp-physics/issues/5
- https://github.com/NOAA-EMC/fv3atm/issues/586

It should also be noted that many users have found dubious values the model receives from the boundary conditions. These bugs may be the cause. However, we'll have to run parallels of multiple configurations to confirm that.

**To Reproduce**

1. Read the code and find the ps_reg=-9999999
2. Replace that with ps_reg=real_snan
3. Compile in DEBUG=ON mode.
4. Run it and see the bugs.
6. Fix the bugs one by one and see more appear.

**Expected behavior**

1. Boundary data should be correct.
2. The boundary code should not write to non-boundary regions within the domain.
3. Blending code should not blend data beyond the domain boundary.
4. The best practice of initializing to real_snan should be followed.
5. The code should pass in DEBUG=ON
8. Changing the threads, layout, or task count, should not change the results.

**System Environment**
The ufs-weather-model build system targets hera.intel and hera.gnu, which use:

- OS: CentOS 7.9.2009
- Compiler(s): Intel 2022.1.2 and GCC 9.2.0
- MPI type, and version: Intel MPI using the SLURM srun launcher
- netCDF Version: 4.7.4
- Configure options: cmake option "-DDEBUG=ON"

**Additional context**
None.

### PR DESCRIPTION
**Pull request by SamuelTrahanNOAA**
**Description**

This removes the workaround of initializing ps_reg to -9999999 Pa in fv_regional_bc.F90 and fixes many other bugs that caused or masked. One bug related to blending has a workaround instead of a fix. A lot of these bugs were fixed by merging in #220. That PR from @kaiyuan-cheng has an elegant replacement for some of my earlier kludgy fixes. Unfortunately, that PR does not allow the model to run in debug mode. I believe this PR has the minimal set of fixes and workarounds needed, until someone comes up with a more elegant solution to the blending bug.

**What is Fixed?**

This PR fixes everything listed here, while related PRs (listed below) at higher levels fix several other issues. See the next section for the one bug that has a workaround instead of a fix.

1. Only valid boundary values are copied to Atm%ps (not to data within the domain)
2. Many uninitialized variables are now initialized with signalling NaN (real_snan)
3. ps_reg is no longer initialzied to -9999999 Pa
4. Replaced "go to 123" with "exit"
5. PR #220 is included. Indexes are corrected so that wind remapping never accesses data outside boundary regions.

**Workaround for Blending Code Receiving Data Outside the Boundary**

The blending code's `array` array contains NaNs in regions it is asked to blend. If my reading of the code is correct, the `array` array is what is being blended, not the boundary conditions themselves. This is probably (but not certainly) regions beyond the boundary in the domain. Any operations with a NaN return NaN (unless the NaN signals), so processing those locations is a no-op (or a SIGFPE). The workaround is to detect non-finite values using `is_not_finite()` and skip them.

I'm concerned by this bug, as it means something may be wrong elsewhere. Someone with better knowledge of the code should look into this. The trouble may originate from the model level, so I've made an issue there (https://github.com/NOAA-EMC/fv3atm/issues/586). It may be as simple as the blending code looping one point farther than it should.

**Testing:**

Fixes #218 

The RRFS regression tests in the ufs-weather-model were compiled in DEBUG=ON mode, and extensively tested on Hera with the Gnu and Intel compilers. Many hours of debugging with gdb, ncview, and other tools. Testing this also required fixing a numerical precision issue in one ccpp physics scheme (https://github.com/ufs-community/ccpp-physics/issues/5, https://github.com/ufs-community/ccpp-physics/pull/6) and changes to the regression tests to get them to fit in 30 minutes (https://github.com/ufs-community/ufs-weather-model/pull/1437). Ultimately, this fixes half of the test variants of RRFS that were failing before (https://github.com/ufs-community/ufs-weather-model/issues/1222). It also obviates the problem of not having enough PS boundary data (https://github.com/ufs-community/ufs-weather-model/issues/1436).

Preliminary results from a real-time parallel have shown this has a small impact on the bulk 1D statistics in a large domain with a large number of ranks. (Hence, boundary ranks have a small fraction of the domain.) However, regression tests have shown it does change the fields. Nobody has done any analysis between those two extremes.

**Checklist:**

Please check all whether they apply or not
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by SamuelTrahanNOAA** [comment]
These changes are going into the RRFS as an emergency fix, in this PR:

https://github.com/NOAA-GSL/GFDL_atmos_cubed_sphere/pull/12

It's headed to the RRFS_dev branch in the NOAA-GSL fork, which is used as the authoritative repo for RRFS parallels. I'm hoping to un-fork the GFDL_atmos_cubed_sphere as soon as possible. The developers do not want wait the several weeks it'll take to update RRFS_dev to the authoritative branches. They want this fix as soon as possible, due to the severity of the bug and its widespread implications.

**Comment by junwang-noaa** [comment]
@yangfanglin Could we ask Kate or Kevin to review this PR? Thanks

**Comment by laurenchilutti** [comment]
Thank you Jun, I wasn't sure who to ask to review this.  I can update the reviewers if you would like.

**Comment by junwang-noaa** [comment]
Thanks, Lauren. I think some dynamics export on UFS regional application can review the PR.

**Comment by kaiyuan-cheng** [comment]
The original code looks good to me. The proposed change does not seem correct because it remaps the winds at cell edges based on surface pressure at the cell center. FV3 needs winds on the cell edges (C- and D-grids). When the model runs in the regional mode, the remapping of the winds in the boundary domain requires the surface pressure to be on the cell edges, which is obtained by the linear interpolation of the surface pressure at the cell center. For example, 0.5*(psc(i,j-1)+psc(i,j)), as shown in the original code. I am uncertain if there is an issue with the boundary condition files. There are four halos of data, which should be more than enough to construct winds at the edges of three halos required by the FV3 algorithm. For this reason, I don’t see when the lack of surface pressure for remapping winds will happen.

File: model/fv_regional_bc.F90 [hunk 1]
```diff
@@ -3975,12 +4214,12 @@ subroutine remap_dwinds_regional_bc(Atm                              &
 
      do k=1,km+1
         do i=is_v,ie_v
-           pe0(i,k) = ak0(k) + bk0(k)*0.5*(psc(i-1,j)+psc(i,j))
+           pe0(i,k) = ak0(k) + bk0(k)*psc(i,j)
```
Comment (bensonr) [inline model/fv_regional_bc.F90:4217]: I have the same question as @kaiyuan-cheng - why are you removing the average ,which provides the pressure at the cell edge values where the winds are defined, in favor of the cell mean value defined at the cell center?

File: model/fv_regional_bc.F90 [hunk 2]
```diff
@@ -3975,12 +4214,12 @@ subroutine remap_dwinds_regional_bc(Atm                              &
...
         enddo
      enddo
      do k=1,npz+1
         do i=is_v,ie_v
-           pe1(i,k) = Atm%ak(k) + Atm%bk(k)*0.5*(psc(i-1,j)+psc(i,j))
+           pe1(i,k) = Atm%ak(k) + Atm%bk(k)*psc(i,j)
```
Comment (bensonr) [inline model/fv_regional_bc.F90:4222]: Same as the previous comment.

File: model/fv_regional_bc.F90 (hunk 1 above)
Comment (SamuelTrahanNOAA) [reply model/fv_regional_bc.F90:4217]: The i-1 value is outside the boundary in some places.

File: model/fv_regional_bc.F90 (hunk 2 above)
Comment (SamuelTrahanNOAA) [reply model/fv_regional_bc.F90:4222]: The j-1 value is outside the boundary in some places.

**Comment by SamuelTrahanNOAA** [comment]
> For this reason, I don’t see when the lack of surface pressure for remapping winds will happen.

Turn back on ps_reg=real_snan and you will see real_snan in those two lines and quite a few others. The code extensively uses uninitialized memory.

**Comment by SamuelTrahanNOAA** [comment]
There are three reasons I know of why surface pressure is unavailable in the wind remapping:

1. PS is unavailable in the file on some velocity points (it is only available on scalar value points).
2. The boundaries are processed in the wrong order, so the j-1 points are unavailable.
3. Some of the i-1 and j-1 locations are outside the boundary region.

However, I don't even need to know that any of those are true, to know the code reads uninitialized areas. Just turn on the safeguard ps_reg=real_snan, and you'll see it fail all over the place.

**Comment by SamuelTrahanNOAA** [comment]
The only way I see to put the i-1 and j-1 back in, is to use i-1 or j-1 when they are available, but only the local point when they're not. I can use my "is_not_finite" subroutine to determine that.

There are points where i-1 or j-1 are inside the domain, where pressure is unavailable, or beyond the boundary regions, where pressure is unavailable. We'll need PS in the boundary files to extend into those regions. I'll update my PS-related issue accordingly.

Also, the "PS" reader will have to read into the next MPI rank's boundary region, to get i-1 and j-1.

File: model/fv_regional_bc.F90 (hunk 2 above)
Comment (SamuelTrahanNOAA) [reply model/fv_regional_bc.F90:4222]: You will find this same issue for j-1 a few lines up, in two places.

**Comment by SamuelTrahanNOAA** [comment]
As requested by @kaiyuan-cheng and @bensonr the i-1 and j-1 two-point averages are back in the code, and I've used a new workaround:

1. The i-1 and j-1 are used when available
2. No averaging is done with i-1 and j-1 are unavailable
3. Pressure from the nearest scalar point is used when the i,j location is also unavailable.

We cannot fix this properly until https://github.com/ufs-community/ufs-weather-model/issues/1436 is corrected. However, this is a better workaround than using a pressure of -9999999 Pa.

File: model/fv_regional_bc.F90 (hunk 2 above)
Comment (SamuelTrahanNOAA) [reply model/fv_regional_bc.F90]: This error has been corrected, along with the similar errors in three other places.

**Comment by JiliDong-NOAA** [comment]
@MatthewPyle-NOAA could you look into this PR? it will be great if we can figure out if there is any potential issues on blending  NaN values at the lateral boundary.

**Comment by MatthewPyle-NOAA** [comment]
Yes @JiliDong-NOAA , but probably won't have time to look at it until early next week.

**Comment by JacobCarley-NOAA** [comment]
Unsure if helpful, but the limited area model approach/design is covered in [Section 2 and Fig. 1 of this paper](https://doi.org/10.1029/2021MS002483)

And I've attached an old description of how the blending was coded from an internal memo.
[Blending.pdf](https://github.com/NOAA-GFDL/GFDL_atmos_cubed_sphere/files/9659171/Blending.pdf)

**Comment by SamuelTrahanNOAA** [comment]
I see this in the paper:

> because remapping of the wind components requires adjacent pressure values, a fourth column/row of boundary grid cells is required to hold pressure

Some points lack the j-1 or i-1 pressure data. I see one place where that *will* happen, according to the figure1 in the paper: when i-1 or j-1 are inside the domain. 

However, I think some of the boundaries may be missing that fourth column/row as well, once it gets to the fv_regional_bc.F90. When I try to extend the region that PS is read in, the NetCDF library tells me that is beyond the bounds of the array. Either the data isn't there, or there's some indexing oddity for PS that the code isn't accounting for.

**Comment by SamuelTrahanNOAA** [comment]
@jkbk2004 I'd like to delay this PR until it has been properly reviewed. It is https://github.com/ufs-community/ufs-weather-model/pull/1437 in the ufs-weather-model repository. Could you please reschedule that PR to next week?

**Review by kaiyuan-cheng — COMMENTED** [review]
I went through the code carefully. It turns out that remap_dwinds_regional_bc() does access uninitialized surface pressure, as @SamuelTrahanNOAA pointed out. It means that winds at certain columns and rows are incorrectly remapped. The good news is that those bad winds are never used by the FV3 dycore. (It is not surprising as we don't observe obvious noise near the boundary.) Although this issue does not affect the solution, a fix would be still desirable. I have the fix ready and I will open a PR shortly.

**Comment by SamuelTrahanNOAA** [comment]
Everyone is hyperfocusing on four lines of code when there are quite a few other bugs in the fv_regional_bc.F90. In particular, the bug causing the model to abort in debug mode is NOT the one involving accessing i-1 and j-1. Since some people don't want to read my bug reports or pull request, I'll give you a hint: -9999999 Pa does propagate up to the model, inside the domain, not just at the boundary.

**Comment by kaiyuan-cheng** [comment]
I do like the idea "ps_reg = real_snan" and I have it in my fix. It does not signal any error in my test case. I am not sure about what other bugs you are referring to.

**Comment by SamuelTrahanNOAA** [comment]
> I do like the idea "ps_reg = real_snan" and I have it in my fix. It does not signal any error in my test case. I am not sure about what other bugs you are referring to.

Good to hear. I will have to see your changes to know what (if anything) is not fixed.

**Comment by SamuelTrahanNOAA** [comment]
@kaiyuan-cheng 's PR is here: #220, but it is on top of main rather than dev/emc. I lack the ability to test this in UFS since it won't compile within UFS. (CMakeFile is missing.) That means I can't even confirm #220 will fix the bugs. It also leaves me wondering if some of the bugs originate from elsewhere in dev/emc, rather than the three boundary index lines fixed in #220. Without the ability to test, I do not know.

**Comment by SamuelTrahanNOAA** [comment]
Although #220 does not fix the problem, I was able to dramatically reduce the complexity of my PR by replacing nearly all of my PS-related fixes with @kaiyuan-cheng's change. I believe this is the minimal possible set of fixes until someone has a more elegant fix to the blending bug.

**Comment by BrianCurtis-NOAA** [comment]
The UFSWM PR that contains this is on the top of the commit queue. I recommend we get reviews/approvals here that there are no visible issues in the code **before** we start testing in UFSWM.

**Comment by SamuelTrahanNOAA** [comment]
I concur with @BrianCurtis-NOAA. I'm less worried about the state of this PR after the fixes from #220 were merged, but I do want someone to vet the remaining changes before widespread testing.

File: model/fv_regional_bc.F90 [hunk 3]
```diff
@@ -3594,10 +3663,10 @@ subroutine remap_scalar_nggps_regional_bc(Atm                         &
...
 !---------------------------------------------------------------------------------
 !
-      is=lbound(Atm%ps,1)
-      ie=ubound(Atm%ps,1)
-      js=lbound(Atm%ps,2)
-      je=ubound(Atm%ps,2)
+      is=min(ubound(Atm%ps,1),max(lbound(Atm%ps,1),is))
+      ie=min(ubound(Atm%ps,1),max(lbound(Atm%ps,1),ie))
+      js=min(ubound(Atm%ps,2),max(lbound(Atm%ps,2),js))
+      je=min(ubound(Atm%ps,2),max(lbound(Atm%ps,2),je))
```
Comment (kaiyuan-cheng) [inline model/fv_regional_bc.F90:3669]: I doubt that the modification here is really necessary.

File: model/fv_regional_bc.F90 [hunk 4]
```diff
@@ -4864,6 +4933,9 @@ subroutine bc_time_interpolation(array                               &
           do j=j1_blend,j2_blend
             factor_dist=exp(-(blend_exp1+blend_exp2*(j-j_bc-1)*rdenom)) !<-- Exponential falloff of blending weights.
             do i=i1_blend,i2_blend
+              if(is_not_finite(array(i,j,k))) then
+                cycle ! Outside boundary
+              endif
```
Comment (kaiyuan-cheng) [inline model/fv_regional_bc.F90:4938]: I don’t run the blending boundary condition so I cannot test it. This walkaround looks fine to me although I do think that revising the indices of the blending domain is the ultimate fix.

Comment (SamuelTrahanNOAA) [reply model/fv_regional_bc.F90:4938]: You're probably right about the indices, but I don't trust my knowledge of the code. If someone else knows the blending code, hopefully they'll review and suggest a better solution.

**Comment by MatthewPyle-NOAA** [comment]
I haven't had time to look in great detail, but philosophically it seems like a bandaid is being applied with these is_not_finite checks.  It would be nice to know the root cause of why it is reaching out beyond defined data.   How frequently are these checks excluding i,j values?              

```
if(is_not_finite(array(i,j,k))) then
                cycle ! Outside boundary
endif
```
The other change that caught my eye was the changes to the regional_bc_bounds variables (for south, north, and west) - I just would be somewhat surprised that it had been wrong all this time given the testing and vetting of the code by the original coders.   But certainly could be possible.

```
-        regional_bc_bounds%is_west_uvw=ied-nhalo_model-nrows_blend+1
+        regional_bc_bounds%is_west_uvw=ied-nhalo_model-nrows_blend+2
```

File: model/fv_regional_bc.F90 (hunk 3 above)
Comment (SamuelTrahanNOAA) [reply model/fv_regional_bc.F90:3669]: Without these changes, the model uses data outside the boundary regions that are overwritten with real_snan in the following loop. This error was how I found the ps_reg=-9999999 and later bugs. (Maybe this is caused by a region of the code you don't typically use?)

However, even if we assume that the changes are not necessary, it is then a harmless bounds check on copying to Atm%ps. As with initializing data to real_snan, bounds checks are reasonable safeguard.

**Comment by SamuelTrahanNOAA** [comment]
@MatthewPyle-NOAA Those index changes come from @kaiyuan-cheng in #220 and they do prevent most out-of-bounds access. However, the results are different than my original fixes in RRFS_dev so these have not been vetted in a parallel yet.

**Comment by SamuelTrahanNOAA** [comment]
> How frequently are these checks excluding i,j values?

I have not tested that, but I recall in earlier testing the first variable that has the problem is "pt". Also, those changes are a no-op; the areas that it is blending have no valid data. 

> I just would be somewhat surprised that it had been wrong all this time given the testing and vetting of the code by the original coders.

Many out-of-bounds bugs were masked by `ps_reg=-9999999` which bypassed the normal safeguard of initializing to signalling NaN. My issue #218 lists the out-of-bounds accesses I found when restoring the safeguard.

**Comment by jkbk2004** [comment]
All tests for ufs-wm [pr #1437](https://github.com/ufs-community/ufs-weather-model/pull/1437) are finished. We can start merging in this pr.

**Comment by jkbk2004** [comment]
@laurenchilutti @SamuelTrahanNOAA @bensonr @kaiyuan-cheng just a reminder to have final approvals for this pr.

**Comment by SamuelTrahanNOAA** [comment]
@jkbk2004 There is nothing I can do to effect that end but wait.

**Review by bensonr — APPROVED** [review]

**Comment by jkbk2004** [comment]
@kaiyuan-cheng It seems Kaiyuan is not available. @laurenchilutti can you review/merge this pr ? commit queue has been on hold to move to next PR.

**Review by kaiyuan-cheng — APPROVED** [review]

**Comment by jkbk2004** [comment]
@laurenchilutti @kaiyuan-cheng @bensonr Thanks for following up with the pr!
